src/python-common/ceph/deployment/service_spec.py

```diff
@@ -3148,5 +3406,44 @@ def to_json(self) -> "OrderedDict[str, Any]":
             ]
         return obj
 
+    def explain() -> Dict:
+        se = SMBSpec.spec_explain
+        se['cluster_id'] = (
+            'A name identifying the smb "cluster" this daemon is part of. '
+            'A cluster may be made up of one or more services sharing a common configuration.'
+        )
+        se['features'] = (
+            'a list of terms enabling specific deployment features. '
+            'Terms include: `domain` to enable Active Dir. Domain membership.'
+        )
+        se['config_uri'] = (
+            'a pseudo-uri that resolves to a configuration source that the '
+            'samba-container can load. A ceph based samba container will be '
+            'typically storing configuration in rados (rados:// prefix)'
+        )
+        se['join_sources'] = (
+            'a list of pseudo-uris that resolve to a (JSON) blob containing '
+            'data the samba-container can use to join a domain. '
+            'A ceph based samba container may typically use a rados uri '
+            'or a mon config-key store uri (example: '
+            '`rados:mon-config-key:smb/config/mycluster/join1.json`).'
+        )
+        se['user_sources'] = (
+            'a list of pseudo-uris that resolve to a (JSON) blob containing '
+            'data the samba-container can use to create users (and/or groups). '
+            'A ceph based samba container may typically use a rados uri '
+            'or a mon config-key store uri (example: '
+            '`rados:mon-config-key:smb/config/mycluster/join1.json`).'
+        )
+        se['custom_dns'] = (
+            'a list of IP addresses that will be set up as '
+            'custom dns servers for the samba container.'
+        )
+        se['include_ceph_users'] = (
+            'A list of ceph auth entity names that will be automatically '
+            'added to the ceph keyring provided to the samba container.'
+        )
+        return se
+
```

Comment:
Sorry, but this is a NACK from me. I do not think this approach is maintainable. You'd be better off trying to add support for (generated?) JSONSchema or something similar to these classes.

Reply:
@phlogistonjohn got your point from the maintainable perspective. Can you provide reference example to make this better ?

The logic I tried to use here is avoiding instantiating of the service class in orchestrator/module.py